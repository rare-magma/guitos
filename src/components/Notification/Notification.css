.toast-container > :not(:last-child) {
  margin-bottom: 0.75rem;
}

.toast {
  background: var(--lightbgcolor);
  border: 1px solid var(--comment);
  color: var(--textcolor);
}
