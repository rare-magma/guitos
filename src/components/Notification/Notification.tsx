import { Button, Toast } from "react-bootstrap";
import { BsX } from "react-icons/bs";

interface NotificationProps {
  notification: { show: boolean; id?: string; body?: string };
  onShow: () => void;
}

function Notification({ notification, onShow }: NotificationProps) {
  return (
    <Toast
      key={`${notification.id}-toast`}
      onClose={() => onShow()}
      show={notification.show}
      autohide
      delay={3000}
    >
      <Toast.Header key={`${notification.id}-toast-header`} closeButton={false}>
        <div
          key={`${notification.id}-toast-body`}
          className="me-auto text-truncate"
          style={{
            textOverflow: "ellipsis",
            overflow: "hidden",
            maxWidth: "90%",
          }}
        >
          {notification.body}
        </div>
        <Button
          key={`${notification.id}-toast-button`}
          size="sm"
          aria-label="dismiss notification"
          variant="outline-secondary"
          onClick={() => onShow()}
        >
          {<BsX />}
        </Button>
      </Toast.Header>
    </Toast>
  );
}
export default Notification;
