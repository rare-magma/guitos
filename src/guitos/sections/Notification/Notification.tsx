import { Button, Toast } from "react-bootstrap";
import { BsArrowCounterclockwise, BsX } from "react-icons/bs";
import { useBudget } from "../../context/BudgetContext";
import "./Notification.css";
import type { BudgetNotification } from "../../context/NotificationContext";

interface NotificationProps {
  notification: BudgetNotification;
  handleClose: (notification: BudgetNotification) => void;
}

export function Notification({ notification, handleClose }: NotificationProps) {
  const { undo } = useBudget();

  return (
    <Toast
      key={`${notification.id}-toast`}
      onClose={() => handleClose(notification)}
      show={notification.show}
      autohide={true}
      delay={notification.showUndo ? 60000 : 3000}
    >
      <Toast.Body
        key={`${notification.id}-toast-body`}
        className="p-2 d-flex justify-content-between align-items-center"
      >
        <div
          key={`${notification.id}-toast-body-div`}
          className="me-2 me-auto text-truncate fw-medium"
          style={{
            textOverflow: "ellipsis",
            overflow: "hidden",
            maxWidth: "90%",
          }}
        >
          {notification.body}
        </div>
        {notification.showUndo && (
          <Button
            className="me-2 align-self-end"
            key={`${notification.id}-undo-button`}
            size="sm"
            aria-label="undo budget deletion"
            variant="outline-info toggle"
            onClick={() => {
              undo();
              handleClose(notification);
            }}
          >
            <BsArrowCounterclockwise aria-hidden={true} />
          </Button>
        )}
        <Button
          className="align-self-end"
          key={`${notification.id}-toast-button`}
          size="sm"
          aria-label="dismiss notification"
          variant="outline-secondary"
          onClick={() => handleClose(notification)}
        >
          {<BsX aria-hidden={true} />}
        </Button>
      </Toast.Body>
    </Toast>
  );
}
