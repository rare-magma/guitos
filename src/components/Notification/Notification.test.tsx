import { render, screen } from "@testing-library/react";
import Notification from "./Notification";
import { vi } from "vitest";
import userEvent from "@testing-library/user-event";

describe("Notification", () => {
  const onShow = vi.fn();

  const notification = {
    show: true,
    id: "a",
    body: "notification body",
  };

  const comp = <Notification notification={notification} onShow={onShow} />;

  beforeEach(() => {
    render(comp);
  });

  it("matches snapshot", () => {
    expect(comp).toMatchSnapshot();
  });

  it("renders initial state", () => {
    expect(screen.getByText("notification body")).toBeInTheDocument();
  });

  it("triggers onShow when closed", async () => {
    await userEvent.click(screen.getByRole("button"));
    expect(onShow).toHaveBeenCalledTimes(1);
  });
});
