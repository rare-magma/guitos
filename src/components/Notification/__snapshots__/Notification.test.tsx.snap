// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Notification > matches snapshot 1`] = `
<Notification
  notification={
    {
      "body": "notification body",
      "id": "a",
      "show": true,
    }
  }
  onShow={[MockFunction spy]}
/>
`;
